xla/service/gpu/gpu_offloading_test.cc

```diff
@@ -132,10 +136,51 @@ ENTRY %main (param_0: f32[1024], param_1: f32[1024]) -> f32[1024] {
   ROOT %res_11 = f32[1024]{0} tanh(f32[1024]{0} %res_10)
 }
 )";
-  EXPECT_TRUE(RunAndCompare(hlo_text, ErrorSpec{/*aabs=*/1e-6, /*arel=*/1e-6}));
+  EXPECT_TRUE(RunAndCompareNoHloPasses(hlo_text, ErrorSpec{1e-3}));
+}
+
+TEST_F(GpuOffloadingTest, FusedComputationOffloadingTest) {
+  const char* hlo_text = R"(
+  HloModule test
+
+  mul {
+    %param_1 = f32[1024]{0} parameter(1)
+    %param_0 = f32[1024]{0} parameter(0)
+    ROOT m = f32[1024]{0} multiply(%param_0, %param_1)
+  }
+
+  exp {
+    %param_0 = f32[1024]{0} parameter(0)
+    e = f32[1024]{0} exponential(%param_0)
+    ROOT t = f32[1024]{0} tanh(e)
+  }
+
+  ENTRY %main (param_0: f32[1024], param_1: f32[1024]) -> f32[1024] {
+  %param_1 = f32[1024]{0} parameter(1)
+  %param_0 = f32[1024]{0} parameter(0)
+  %res_3 = f32[1024]{0} fusion(%param_1, %param_0), kind=kInput, calls=mul
+  %copy-start = (f32[1024]{0:S(5)}, f32[1024]{0}, u32[]) copy-start(f32[1024]{0} %res_3)
```

Comment:
@zhenying-liu What mechanism is used to ensure that for example res_3 result is ready by the time the copy-start is issued on an separate stream? This test fails for rocm and I'm not sure if we are missing something here.

Reply:
@draganmladjenovic That's a good question. I haven't thought about it before. @Tixxx Do you think I should let the other stream wait for res_3 from the main stream before executing the memcpy issued by the copy-start?

Reply:
Yes, we always need to put a stream dependency on the main stream to ensure data is ready, i.e:
`async_stream.WaitFor(params.stream)`

Reply:
Thanks @Tixxx for clarification. @draganmladjenovic Will add ASAP.

Reply:
@draganmladjenovic I revisited the code and the mechanism wasn't missing. res_3 and copy-start are executed on the same stream sequentially so that the order of execution is guaranteed. copy-start will do memcpy d2h after res_3 is defined, and create cuda events. This is done by DeviceToHostCopyThunk. The other stream will be waiting for this event, and the waiting is issued by CopyDoneThunk.

Reply:
> @draganmladjenovic I revisited the code and the mechanism wasn't missing. res_3 and copy-start are executed on the same stream sequentially so that the order of execution is guaranteed. copy-start will do memcpy d2h after res_3 is defined, and create cuda events. This is done by DeviceToHostCopyThunk. The other stream will be waiting for this event, and the waiting is issued by CopyDoneThunk.

Yes. But there seem to be no ordering between d2h and res_3 being written to the memory. My understanding is that copy-start is executed on different stream than main stream.

Reply:
copy-start must be executed on the main stream:
https://github.com/openxla/xla/pull/10782/files#diff-af8b05a6cc3f923b94c21be0b86ea1660e2216200c27d9bddedd0d99855dc3f4R118-R121

Reply:
That is a different stream for me in the test.

Reply:
Yes, you are right. The memcpy d2h is from a different stream. Thanks for pointing out the issue.
I made a [PR draft 13597](https://github.com/openxla/xla/pull/13597) advised by @Tixxx.